api/layer/multipart_upload.go

```diff
@@ -169,6 +171,45 @@ func (n *layer) UploadPartCopy(ctx context.Context, p *UploadCopyParams) (*data.
 	return n.CopyObject(ctx, c)
 }
 
+// implements io.Reader of payloads of the object list stored in the NeoFS network.
+type multiObjectReader struct {
+	ctx context.Context
+
+	layer *layer
+
+	prm getParams
+
+	curReader io.Reader
+
+	parts []*data.ObjectInfo
+}
+
+func (x *multiObjectReader) Read(p []byte) (n int, err error) {
+	if x.curReader != nil {
+		n, err = x.curReader.Read(p)
+		if err == nil || err != nil && !stderrors.Is(err, io.EOF) {
```

Comment:
Condition can be simplified to `if err == nil || !stderrors.Is(err, io.EOF)` or even to `if !stderrors.Is(err, io.EOF)` (though checking to nil is a little optimization)